test/src/main/java/ai/timefold/solver/test/impl/score/stream/AbstractSingleConstraintAssertion.java

```diff
@@ -0,0 +1,745 @@
+package ai.timefold.solver.test.impl.score.stream;
+
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiPredicate;
+import java.util.stream.Stream;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.api.score.constraint.ConstraintMatchTotal;
+import ai.timefold.solver.core.api.score.constraint.Indictment;
+import ai.timefold.solver.core.api.score.stream.ConstraintJustification;
+import ai.timefold.solver.core.impl.score.DefaultScoreExplanation;
+import ai.timefold.solver.core.impl.score.definition.ScoreDefinition;
+import ai.timefold.solver.core.impl.score.stream.common.AbstractConstraint;
+import ai.timefold.solver.core.impl.score.stream.common.AbstractConstraintStreamScoreDirectorFactory;
+import ai.timefold.solver.core.impl.score.stream.common.ScoreImpactType;
+import ai.timefold.solver.core.impl.util.Pair;
+import ai.timefold.solver.test.api.score.stream.SingleConstraintAssertion;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public abstract sealed class AbstractSingleConstraintAssertion<Solution_, Score_ extends Score<Score_>>
+        implements SingleConstraintAssertion permits DefaultSingleConstraintAssertion, DefaultSingleConstraintListener {
+
+    private final AbstractConstraint<Solution_, ?, ?> constraint;
+    private final ScoreDefinition<Score_> scoreDefinition;
+    private Score_ score;
+    private Collection<ConstraintMatchTotal<Score_>> constraintMatchTotalCollection;
+    private Collection<ConstraintJustification> justificationCollection;
+    private Collection<Indictment<Score_>> indictmentCollection;
+
+    @SuppressWarnings("unchecked")
+    AbstractSingleConstraintAssertion(AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_, ?> scoreDirectorFactory) {
+        this.constraint = (AbstractConstraint<Solution_, ?, ?>) scoreDirectorFactory.getConstraintMetaModel()
+                .getConstraints()
+                .stream()
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Impossible state: no constraint found."));
+        this.scoreDefinition = scoreDirectorFactory.getScoreDefinition();
+    }
+
+    @SuppressWarnings("java:S1905")
```

Comment:
Let's not overdo it with Sonar suppressions, it's clutter.
We can get rid of this in Sonar UI, or not at all. (We don't mind minor issues.)